File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-7.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

"""
two passes
first to calculate length
second to remove node
"""

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        length = 0
        secondLast = None
        node = head

        while node.next:
            length += 1
            secondLast = node
            node = node.next
        
        length += 1
        

        targetIndex = length - n

        if targetIndex == 0:
            temp = head.next
            head.next = None

            return temp
        elif n == 1:
            secondLast.next = None
            return head
        

        i = 0
        predecessor = head
        while predecessor:
            if i == targetIndex - 1:
                target = predecessor.next
                successor = target.next
                target.next = None

                predecessor.next = successor
                
                return head
            else:
                predecessor = predecessor.next
                i += 1

File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-7.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        PriorityQueue<ListNode> minHeap = new PriorityQueue<>((a, b) -> Integer.compare(a.val, b.val));

        ListNode head = new ListNode();
        ListNode curr = head;
        for (ListNode node : lists) minHeap.offer(node);

        while (minHeap.size() > 0) {
            ListNode node = minHeap.poll();
            if (node.next != null) minHeap.offer(node.next);

            curr.next = node;
            curr = node;
        }

        return head.next;
    }
}
